File: backend/app/schemas/crowd.py
from typing import List, Optional
from pydantic import BaseModel, Field


class HourlyPrediction(BaseModel):
    time: str
    crowd_percent: int = Field(ge=0, le=100)
    expected_visitors: int = Field(ge=0)


class CrowdPredictionRequest(BaseModel):
    site_id: str
    date: Optional[str] = None
    weather: Optional[str] = None
    temperature: Optional[float] = None


class CrowdPredictionResponse(BaseModel):
    site_id: str
    site_name: str
    city: str
    state: str
    date: str
    day_of_week: str
    operating_hours: str
    weather: str
    temperature_c: float
    safe_capacity: int
    daily_expected_total: int
    predictions: List[HourlyPrediction]
    peak_hours: List[str]
    best_time: str

File: frontend/src/api/crowd.ts
import { API_BASE_URL } from './config';

export interface HourlyPrediction {
  time: string; // "09:00"
  crowd_percent: number; // e.g. 42
  expected_visitors: number; // e.g. 630
}

export interface CrowdPredictionResponse {
  site_id: string;
  site_name: string;
  city: string;
  state: string;
  date: string; // "2026-08-30"
  day_of_week: string; // "Sunday"
  operating_hours: string; // "09:00-16:00"
  weather: string; // "Clear"
  temperature_c: number; // 28
  safe_capacity: number; // 15000
  daily_expected_total: number; // 8400
  predictions: HourlyPrediction[];
  peak_hours: string[]; // ["11:00-14:00"]
  best_time: string; // "09:00-11:00"
}

export async function getCrowd(
  siteId: string,
  date?: string
): Promise<CrowdPredictionResponse> {
  const url = new URL(`${API_BASE_URL}/api/crowd/${siteId}`);
  if (date) {
    url.searchParams.set("date", date);
  }

  const response = await fetch(url.toString());

  if (!response.ok) {
    throw new Error(
      `Failed to fetch crowd for site ${siteId}: ${response.status} ${response.statusText}`
    );
  }

  return response.json();
}